### PR TITLE
Implement endoplasmic reticulum

### PR DESCRIPTION
**Pull request by wizmer**
Provide an API to retrieve data in the /organelles/endoplasmic_reticulum
group.

One can retrieve the following properties:
- neuronal section indices
- volumes for each of the listed neuronal section
- surface ereas for each of the listed neuronal section
- filament counts for each of the listed neuronal section

Spec at:
https://bbpteam.epfl.ch/documentation/projects/Morphology%20Documentation/latest/h5v1.html

### CODE REVIEW

**Comment by wizmer** [comment]
This will be in another commit

File: include/morphio/endoplasmic_reticulum.h [hunk 1]
```diff
@@ -0,0 +1,43 @@
...
+    **/
+    const std::vector<float>& surfaceAreas() const;
+
+    /**
+       Returns the number of filaments for each neuronal section
+    **/
+    const std::vector<uint32_t>& filamentCounts() const;
+
+private:
+    EndoplasmicReticulum(std::shared_ptr<Property::Properties> properties)
```
Comment (mgeplf) [inline include/morphio/endoplasmic_reticulum.h]: I'm still don't fully get why this is private, and then you have to declare `Morphology` a friend.  Seems like you're allowing `Morphology` much more access (ie: to _properties) than is necessary.

File: src/CMakeLists.txt [hunk 2]
```diff
@@ -3,6 +3,7 @@ set(MORPHIO_SOURCES
     errorMessages.cpp
     mito_section.cpp
     mitochondria.cpp
+    endoplasmic_reticulum.cpp
```
Comment (mgeplf) [inline src/CMakeLists.txt]: please keep this sorted

File: src/endoplasmic_reticulum.cpp [hunk 3]
```diff
@@ -0,0 +1,25 @@
+#include <morphio/properties.h>
+#include <morphio/endoplasmic_reticulum.h>
+
+namespace morphio {
+const std::vector<uint32_t>& EndoplasmicReticulum::sectionIndices() const
+{
+    return _properties->_endoplasmicReticulumLevel._sectionIndex;
+}
+
+const std::vector<float>& EndoplasmicReticulum::volumes() const
```
Comment (mgeplf) [inline src/endoplasmic_reticulum.cpp]: how is the lifetime of the vector guaranteed?

File: src/mitochondria.cpp [hunk 4]
```diff
@@ -1,5 +1,6 @@
 #include <morphio/mito_section.h>
 #include <morphio/mitochondria.h>
+#include <morphio/properties.h>
```
Comment (mgeplf) [inline src/mitochondria.cpp]: why is this needed?

File: src/readers/morphologyHDF5.cpp [hunk 5]
```diff
@@ -486,6 +494,29 @@ void MorphologyHDF5::_read(const std::string& groupName,
...
 }
 
+void MorphologyHDF5::_readEndoplasmicReticulum()
+{
+    {
+        HighFive::SilenceHDF5 silence;
+
+        try {
+            const auto group = _file->getGroup(_g_endoplasmic_reticulum);
+        } catch (HighFive::GroupException&) {
```
Comment (mgeplf) [inline src/readers/morphologyHDF5.cpp]: what's the semantics of getting this? can we check if the file has the property first, instead of catching the (potentially expensive) exception?

File: include/morphio/endoplasmic_reticulum.h (hunk 1 above)
Comment (wizmer) [reply include/morphio/endoplasmic_reticulum.h]: > I'm still don't fully get why this is private, and then you have to declare Morphology a friend. 

My idea was that in MorphIO the *only* public constructor would be `Morphology`.
I don't want people to start creating EndoplasmicReticulum, or Section, Mitochondria objects that are not attached to a `Morphology` object.
You should only access those objects through `Morphology`:
```
m = Morphology("my-morph.swc");
m.mitochondria().do_stuff();
m.section(1).do_other_stuff();
m.endoplasmicReticulum().do_even_more_stuff();
```

> Seems like you're allowing Morphology much more access (ie: to _properties) than is necessary.

She only place where the EndoplasmicReticulum is called is in `Morphology::endoplasmicReticulum()` and the passed `_properties` structure is already owned by `Morphology`

Does this make sense ?

File: src/CMakeLists.txt (hunk 2 above)
Comment (wizmer) [reply src/CMakeLists.txt]: good point

File: src/readers/morphologyHDF5.cpp (hunk 5 above)
Comment (wizmer) [reply src/readers/morphologyHDF5.cpp]: This semantics comes straight from Brion. I can have a deeper look if you want.

Comment (mgeplf) [reply src/readers/morphologyHDF5.cpp]: Nope, it's ok.

Comment (tristan0x) [reply src/readers/morphologyHDF5.cpp]: There is a `NodeTraits::exist` method that does what you need:
https://github.com/BlueBrain/HighFive/blob/74250cf2f2a82a2b9af2f00f93475c431a2d7e09/include/highfive/bits/H5Node_traits.hpp#L110-L115

Otherwise I would consider:
```suggestion
        } catch (const HighFive::GroupException&) {
```

File: include/morphio/endoplasmic_reticulum.h [hunk 6]
```diff
@@ -0,0 +1,43 @@
...
+        : _properties(properties)
```
Comment (tristan0x) [inline include/morphio/endoplasmic_reticulum.h]: To save an atomic increment in `_ptr` constructor and an atomic decrement in `properties` destructor:

```suggestion
        : _properties(std::move(properties))
```

Comment (wizmer) [reply include/morphio/endoplasmic_reticulum.h]: Good point. Thanks for the suggestion.

File: src/endoplasmic_reticulum.cpp (hunk 3 above)
Comment (wizmer) [reply src/endoplasmic_reticulum.cpp]: The lifetime will be the one of the Morphology object.

File: src/mitochondria.cpp (hunk 4 above)
Comment (wizmer) [reply src/mitochondria.cpp]: Not needed, indeed.

File: src/endoplasmic_reticulum.cpp [hunk 7]
```diff
@@ -0,0 +1,25 @@
+#include <morphio/properties.h>
+#include <morphio/endoplasmic_reticulum.h>
+
+namespace morphio {
+const std::vector<uint32_t>& EndoplasmicReticulum::sectionIndices() const
```
Comment (tristan0x) [inline src/endoplasmic_reticulum.cpp]:
```suggestion
const std::vector<uint32_t>& EndoplasmicReticulum::sectionIndices() const noexcept
```

File: src/endoplasmic_reticulum.cpp (hunk 3 above)
Comment (tristan0x) [inline src/endoplasmic_reticulum.cpp]:
```suggestion
const std::vector<float>& EndoplasmicReticulum::volumes() const noexcept
```

File: src/endoplasmic_reticulum.cpp [hunk 8]
```diff
@@ -0,0 +1,25 @@
...
+{
+    return _properties->_endoplasmicReticulumLevel._sectionIndex;
+}
+
+const std::vector<float>& EndoplasmicReticulum::volumes() const
+{
+    return _properties->_endoplasmicReticulumLevel._volume;
+}
+
+const std::vector<float>& EndoplasmicReticulum::surfaceAreas() const
```
Comment (tristan0x) [inline src/endoplasmic_reticulum.cpp]:
```suggestion
const std::vector<float>& EndoplasmicReticulum::surfaceAreas() const noexcept
```

File: src/endoplasmic_reticulum.cpp [hunk 9]
```diff
@@ -0,0 +1,25 @@
...
+{
+    return _properties->_endoplasmicReticulumLevel._surfaceArea;
+}
+
+const std::vector<uint32_t>& EndoplasmicReticulum::filamentCounts() const
```
Comment (tristan0x) [inline src/endoplasmic_reticulum.cpp]:
```suggestion
const std::vector<uint32_t>& EndoplasmicReticulum::filamentCounts() const noexcept
```

File: src/morphology.cpp [hunk 10]
```diff
@@ -87,6 +88,11 @@ Mitochondria Morphology::mitochondria() const
     return Mitochondria(_properties);
 }
 
+const EndoplasmicReticulum Morphology::endoplasmicReticulum() const
```
Comment (tristan0x) [inline src/morphology.cpp]:
```suggestion
EndoplasmicReticulum Morphology::endoplasmicReticulum() const
```

File: src/readers/morphologyHDF5.cpp [hunk 11]
```diff
@@ -479,6 +487,29 @@ void MorphologyHDF5::_read(const std::string& groupName,
...
 
+void MorphologyHDF5::_readEndoplasmicReticulum()
+{
+    {
+        HighFive::SilenceHDF5 silence;
+
+        try {
+            const auto group = _file->getGroup(_g_endoplasmic_reticulum);
+        } catch (const HighFive::GroupException&) {
+            return;
```
Comment (tristan0x) [inline src/readers/morphologyHDF5.cpp]: Don't we want to emit an error message here?

File: include/morphio/endoplasmic_reticulum.h (hunk 1 above)
Comment (mgeplf) [reply include/morphio/endoplasmic_reticulum.h]: Ok, I think I get it

File: src/readers/morphologyHDF5.cpp (hunk 11 above)
Comment (wizmer) [reply src/readers/morphologyHDF5.cpp]: No. The endoplasmic group is optional.
